File: backend/internal/service/net.go
package service

import (
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"sync"

	"CorrectQuiz.com/quiz/internal/entity"
	"github.com/gofiber/contrib/websocket"
	"github.com/google/uuid"
)

type NetService struct {
	quizService *QuizService
	games       []*Game
	gamesMutex  sync.RWMutex
}

func Net(quizService *QuizService) *NetService {
	return &NetService{
		quizService: quizService,
		games:       []*Game{},
		gamesMutex:  sync.RWMutex{},
	}
}

type ConnectPacket struct {
	Code string `json:"code"`
	Name string `json:"name"`
}

type HostGamePacket struct {
	QuizId string `json:"quizId"`
}

type QuestionShowPacket struct {
	Question      entity.QuizQuestion `json:"question"`
	QuestionIndex int                 `json:"questionIndex"`
}

type ChangeGameStatePacket struct {
	State GameState `json:"state"`
	Code  string    `json:"code,omitempty"`
}

type PlayerJoinPacket struct {
	Player Player `json:"player"`
}

type StartGamePacket struct {
}

type TickPacket struct {
	Tick int `json:"tick"`
}

type QuestionAnswerPacket struct {
	Question int `json:"question"`
	Choice   int `json:"choice"`
}

type PlayerRevealPacket struct {
	Points int `json:"points"`
}

type QuestionRevealPacket struct {
	Question           entity.QuizQuestion `json:"question"`
	CorrectAnswerIndex []int               `json:"correctAnswerIndex"`
	AnswerCounts       []int               `json:"answerCounts"`
}

type LeaderboardPacket struct {
	Points []LeaderboardEntry `json:"points"`
}

type PlayerLeavePacket struct {
	PlayerId uuid.UUID `json:"playerId"`
}

type PlayerRankPacket struct {
	Rank int `json:"rank"`
}

type KickPlayerPacket struct {
	PlayerId string `json:"playerId"`
}

type HostLeavePacket struct{}

type NextQuestionPacket struct{}

func (c *NetService) packetIdtoPacket(packetId uint8) any {
	switch packetId {
	case 0:
		{
			return &ConnectPacket{}
		}
	case 1:
		{
			return &HostGamePacket{}
		}
	case 3:
		{
			return &ChangeGameStatePacket{}
		}
	case 5:
		{
			return &StartGamePacket{}
		}
	case 7:
		{
			return &QuestionAnswerPacket{}
		}
	case 8:
		{
			return &PlayerAnswerFeedbackPacket{
				IsCorrect:          false,
				CorrectAnswerIndex: nil,
			}
		}
	case 9:
		{
			return &AnswerReceivedPacket{}
		}
	case 13:
		{
			return &NextQuestionPacket{}
		}
	case 14:
		{
			return &PlayerRankPacket{}
		}
	case 15:
		{
			return &KickPlayerPacket{}
		}
	case 16:
		{
			return &HostLeavePacket{}
		}
	case 17:
		{
			return &PlayerLeavePacket{}
		}
	}

	return nil
}

func (c *NetService) packettoPacketId(packet any) (uint8, error) {
	switch packet.(type) {
	case HostGamePacket:
		{
			return 1, nil
		}
	case QuestionShowPacket:
		{
			return 2, nil
		}
	case ChangeGameStatePacket:
		{
			return 3, nil
		}
	case PlayerJoinPacket:
		{
			return 4, nil
		}
	case TickPacket:
		{
			return 6, nil
		}
	case PlayerAnswerFeedbackPacket:
		{
			return 8, nil
		}
	case QuestionRevealPacket:
		{
			return 10, nil
		}
	case PlayerRevealPacket:
		{
			return 11, nil
		}
	case LeaderboardPacket:
		{
			return 12, nil
		}
	case PlayerRankPacket:
		{
			return 14, nil
		}
	case *HostLeavePacket:
		{
			return 16, nil
		}
	case PlayerLeavePacket:
		{
			return 17, nil
		}
	}
	return 0, errors.New("invalid packet type")
}

func (c *NetService) GetGameByCode(code string) *Game {
	c.gamesMutex.RLock()
	defer c.gamesMutex.RUnlock()
	for _, game := range c.games {
		if game.Code == code {
			return game
		}
	}
	return nil
}

func (c *NetService) GetGameByHost(host *websocket.Conn) *Game {
	c.gamesMutex.RLock()
	defer c.gamesMutex.RUnlock()
	for _, game := range c.games {
		if game.Host == host {
			return game
		}
	}
	return nil
}

func (c *NetService) GetGameByPlayer(con *websocket.Conn) (*Game, *Player) {
	c.gamesMutex.RLock()
	defer c.gamesMutex.RUnlock()
	for _, game := range c.games {
		game.playersMutex.RLock()
		for _, player := range game.Players {
			if player.Connection == con {
				game.playersMutex.RUnlock()
				return game, player
			}
		}
		game.playersMutex.RUnlock()
	}
	return nil, nil
}

func (c *NetService) IsNameTakenInGame(code string, name string) bool {

	game := c.GetGameByCode(code)
	if game == nil {
		return false
	}

	game.playersMutex.RLock()
	defer game.playersMutex.RUnlock()

	for _, player := range game.Players {
		if player.Name == name {
			return true
		}
	}

	return false
}

func (c *NetService) FindActiveGameByCode(code string) (*Game, error) {
	c.gamesMutex.RLock()
	defer c.gamesMutex.RUnlock()

	for _, game := range c.games {
		if game.Code == code {
			return game, nil
		}
	}
	return nil, errors.New("active game not found")
}

func (c *NetService) handleHostLeave(con *websocket.Conn) {
	game := c.GetGameByHost(con)

	if game != nil {
		endPacket := ChangeGameStatePacket{
			State: GameEndedState,
		}

		if game.cancelFunc != nil {
			game.cancelFunc()
		}

		for _, p := range game.Players {
			if p.Connection != nil {
				err := c.SendPacket(p.Connection, endPacket)
				if err != nil {
				}

				p.Connection.Close()
			}
		}
		c.gamesMutex.Lock()
		var newGames []*Game
		for _, g := range c.games {
			if g.Id != game.Id {
				newGames = append(newGames, g)
			}
		}
		c.games = newGames
		c.gamesMutex.Unlock()
	}
}

func (c *NetService) handlePlayerLeave(con *websocket.Conn, playerId uuid.UUID) {
	game, player := c.GetGameByPlayerId(playerId)
	if game != nil && player != nil {

		game.RemovePlayer(player.Id)

		leavePacket := PlayerLeavePacket{
			PlayerId: player.Id,
		}
		err := c.SendPacket(game.Host, leavePacket)
		if err != nil {
		}

	}
}

func (c *NetService) OnIncomingMessage(con *websocket.Conn, mt int, msg []byte) {

	if len(msg) < 2 {
		return
	}

	packetId := msg[0]
	data := msg[1:]

	packet := c.packetIdtoPacket(packetId)
	if packet == nil {
		return
	}

	err := json.Unmarshal(data, packet)
	if err != nil {
		fmt.Println(err)
		return
	}

	switch data := packet.(type) {
	case *ConnectPacket:
		{
			game := c.GetGameByCode(data.Code)
			if game == nil {
				return
			}
			game.OnPlayerJoin(data.Name, con)
			break
		}

	case *HostGamePacket:
		{

			id64, err := strconv.ParseUint(data.QuizId, 10, 64)
			if err != nil {
				return
			}
			quizId := uint(id64)

			quiz, err := c.quizService.GetQuizById(quizId)
			if err != nil {
				fmt.Println(err)
				return
			}

			if quiz == nil {
				fmt.Println("Quiz not found (ID:", quizId, ")")
				return
			}

			game := newGame(*quiz, con, c)
			c.gamesMutex.Lock()
			c.games = append(c.games, &game)
			c.gamesMutex.Unlock()

			c.SendPacket(con, HostGamePacket{
				QuizId: game.Code,
			})

			c.SendPacket(con, ChangeGameStatePacket{
				State: game.State,
				Code:  game.Code,
			})
			break
		}

	case *StartGamePacket:
		{
			game := c.GetGameByHost(con)
			if game == nil {
				return
			}
			game.Start()
			break
		}
	case *KickPlayerPacket:
		{
			game := c.GetGameByHost(con)
			if game == nil {
				return
			}

			var playerToKick *Player = nil
			for _, player := range game.Players {
				if player.Id.String() == data.PlayerId {
					playerToKick = player
					break
				}
			}

			if playerToKick == nil {
				return
			}

			game.KickPlayer(data.PlayerId)

			break
		}
	case *QuestionAnswerPacket:
		{
			game, player := c.GetGameByPlayer(con)
			if game == nil {
				return
			}

			game.OnPlayerAnswer(data.Question, data.Choice, player)
			break

		}

	case *ChangeGameStatePacket:
		{
			game := c.GetGameByHost(con)
			if game == nil {
				return
			}

			if data.State == IntermissionState {
				game.Intermission()
			}
			break
		}
	case *NextQuestionPacket:
		{
			game := c.GetGameByHost(con)
			if game == nil {
				return
			}
			game.NextQuestion()
			break
		}
	case *PlayerLeavePacket:
		{
			c.handlePlayerLeave(con, data.PlayerId)
			break
		}
	case *HostLeavePacket:
		{
			c.handleHostLeave(con)
			break
		}
	}
}

func (c *NetService) GetGameByPlayerId(playerId uuid.UUID) (*Game, *Player) {
	c.gamesMutex.RLock()
	defer c.gamesMutex.RUnlock()

	for _, game := range c.games {
		game.playersMutex.RLock()
		for _, player := range game.Players {
			if player.Id == playerId {
				game.playersMutex.RUnlock()
				return game, player
			}
		}
		game.playersMutex.RUnlock()
	}
	return nil, nil
}

func (c *NetService) OnDisconnect(con *websocket.Conn) {
	game, player := c.GetGameByPlayer(con)
	if game != nil && player != nil {

		game.RemovePlayer(player.Id)
		leavePacket := PlayerLeavePacket{PlayerId: player.Id}

		err := c.SendPacket(game.Host, leavePacket)
		if err != nil {
		}
		return
	}

	game = c.GetGameByHost(con)
	if game != nil {

		if game.cancelFunc != nil {
			game.cancelFunc()
		}

		endPacket := ChangeGameStatePacket{
			State: GameEndedState,
		}

		game.playersMutex.RLock()
		for _, p := range game.Players {
			if p.Connection != nil {
				c.SendPacket(p.Connection, endPacket)
				p.Connection.Close()
			}
		}
		game.playersMutex.RUnlock()

		c.gamesMutex.Lock()
		var newGames []*Game

		for _, g := range c.games {
			if g.Code != game.Code {
				newGames = append(newGames, g)
			}
		}
		c.games = newGames
		c.gamesMutex.Unlock()

		return
	}
}

func (c *NetService) SendPacket(connection *websocket.Conn, packet any) error {
	bytes, err := c.PacketToBytes(packet)
	if err != nil {
		return err
	}
	return connection.WriteMessage(websocket.BinaryMessage, bytes)
}

func (c *NetService) PacketToBytes(packet any) ([]byte, error) {
	packetId, err := c.packettoPacketId(packet)
	if err != nil {
		return nil, err
	}

	bytes, err := json.Marshal(packet)
	if err != nil {
		return nil, err
	}

	final := append([]byte{packetId}, bytes...)
	return final, nil
}

File: frontend/src/service/net.ts
import { Player, QuizQuestion } from "../model/quiz";

export enum PacketTypes {
    Connect = 0,
    HostGame = 1,
    StartGame = 5,
    Answer = 7,
    AnswerReceived = 9,
    QuestionShow = 2,
    ChangeGameState = 3,
    PlayerJoin = 4,
    Tick = 6,
    PlayerAnswerFeedback = 8,
    QuestionReveal = 10,
    PlayerReveal = 11,
    Leaderboard = 12,
    NextQuestion = 13,
    PlayerRank = 14,
    KickPlayer = 15,
    HostLeave = 16,
    PlayerLeave = 17,
}

export enum GameState {
    Lobby,
    Play,
    Intermission,
    Reveal,
    End,
    GameEndedState,
}

export interface Packet {
    id: PacketTypes;
}

export interface KickPlayerPacket extends Packet {
    playerId: string;
}

export interface HostGamePacket extends Packet {
    quizId: string;
}

export interface ChangeGameStatePacket extends Packet {
    state: GameState;
    code?: string;
}
export interface PlayerJoinPacket extends Packet {
    player: Player;
}

export interface TickPacket extends Packet {
    tick: number;
}

export interface ConnectPacket extends Packet {
    code: string;
    name: string;
}

export interface QuestionShowPacket extends Packet {
    question: QuizQuestion;
    questionIndex: number;
}

export interface QuestionAnswerPacket extends Packet {
    question: number;
    choice: number;
}

export interface PlayerRevealPacket extends Packet {
    points: number;
}

export interface PlayerAnswerFeedbackPacket extends Packet {
    correctAnswerIndex: number[];
    isCorrect: boolean;
    streakBonus: number;
    maxStreak: number;
}

export interface QuestionRevealPacket extends Packet {
    question: QuizQuestion;
    correctAnswerIndex: number[];
    answerCounts: number[];
    maxStreak: number;
}

export interface LeaderboardEntry {
    name: string;
    points: number;
    correctCount: number;
}

export interface LeaderboardPacket extends Packet {
    points: LeaderboardEntry[];
}

export interface PlayerRankPacket extends Packet {
    rank: number;
}

export interface NextQuestionPacket extends Packet { }

export interface PlayerLeavePacket extends Packet {
    playerId: string;
}


export class NetService {
    private webSocket!: WebSocket;
    private textDecoder: TextDecoder = new TextDecoder();
    private textEncoder: TextEncoder = new TextEncoder();

    private onPacketCallback?: (packet: any) => void;
    public onDisconnectCallback?: (code: number, reason: string) => void

    private pendingQueue: Uint8Array[] = [];

    constructor() {
        console.log("DEBUG: NetService initialized. Waiting for connect() call.");
    }
    

    connect(token: string) {
        let baseUrl = (import.meta as any).env.VITE_WS_URL || "ws://localhost:3000/ws";
        const WS_URL = `${baseUrl}?token=${token}`;
        console.log("📢 WS URL is:", WS_URL);
        console.log("📢 Connecting WS with token:", WS_URL);
        this.webSocket = new WebSocket(WS_URL);
        this.webSocket.onopen = () => {
            console.log("opened connection");
            while (this.pendingQueue.length > 0) {
                const data = this.pendingQueue.shift();
                if (data) {
                    console.log("🚀 Sending queued packet...");
                    this.webSocket.send(data);
                }
            }
        };

        this.webSocket.onmessage = async (event: MessageEvent) => {
            const arrayBuffer = await event.data.arrayBuffer();
            const bytes = new Uint8Array(arrayBuffer);
            const packetId = bytes[0];

            const packet = JSON.parse(this.textDecoder.decode(bytes.subarray(1)));

            packet.id = packetId;

            if (this.onPacketCallback)
                this.onPacketCallback(packet);
        }

        this.webSocket.onclose = (event) => {
            if (this.onDisconnectCallback) {
                this.onDisconnectCallback(event.code, event.reason);
            }
        };
    }

    public onDisconnect(callback: (code: number, reason: string) => void) {
        this.onDisconnectCallback = callback;
    }

    public disconnect() {
        if (this.webSocket && this.webSocket.readyState === WebSocket.OPEN) {
            this.webSocket.close();
            console.log("WebSocket connection closed.");
        }
    }



    onPacket(callback: (packet: Packet) => void) {
        this.onPacketCallback = callback;
    }

    sendPacket(packet: Packet) {
        if (!this.webSocket || this.webSocket.readyState !== WebSocket.OPEN) {
            console.warn("⚠️ Socket not ready. Skipping packet:", packet.id);
            return;
        }
        const packetId = packet.id;
        const packetData = JSON.stringify(packet, (key, value) =>
            key == "id" ? undefined : value
        );

        const packetIdArray = new Uint8Array([packetId]);
        const packetDataArray = this.textEncoder.encode(packetData);

        const mergedArray = new Uint8Array(
            packetIdArray.length + packetDataArray.length,
        );
        mergedArray.set(packetIdArray);
        mergedArray.set(packetDataArray, packetIdArray.length);

        if (this.webSocket && this.webSocket.readyState === WebSocket.OPEN) {
            this.webSocket.send(mergedArray);
        } else {
            console.log("⏳ Socket not ready. Queuing packet:", packet.id);
            this.pendingQueue.push(mergedArray);
        }

        this.webSocket.send(mergedArray);
    }
}